### PR TITLE
GSW-1838 refactor: pool's position

### PR DESCRIPTION
**Pull request by r3v4s**
## refactor
1. pool's position
- clearify variable name
- add `position_test.gno` for `position.gno`

### CODE REVIEW

File: pool/position.gno [hunk 1]
```diff
@@ -26,11 +44,11 @@ func positionGetKey(
 // positionUpdateWithKey updates a position in the pool and returns the updated position.
 func (pool *Pool) positionUpdateWithKey(
```
Comment (dongwon8247) [inline pool/position.gno:45]: didn't we agree on changing this method to something else? I don't quite remember but I think we did talk about this

File: pool/position_test.gno [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	for _, tc := range testStruct {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { positionGetKey(tc.owner, tc.tickLower, tc.tickUpper) })
+		} else {
+			uassert.NotPanics(t, func() { positionGetKey(tc.owner, tc.tickLower, tc.tickUpper) })
+		}
+	}
+}
+
+func Test_positionUpdateWithKey(t *testing.T) {
```
Comment (dongwon8247) [inline pool/position_test.gno:43]:
```suggestion
func TestPositionUpdateWithKey(t *testing.T) {
```

File: pool/position.gno [hunk 3]
```diff
@@ -44,49 +62,45 @@ func (pool *Pool) positionUpdateWithKey(
 
 // positionUpdate calculates and returns an updated PositionInfo.
 func positionUpdate(
```
Comment (dongwon8247) [inline pool/position.gno:64]: Can you add a unit test for this as well

**Review by dongwon8247 — CHANGES_REQUESTED** [review]

File: pool/position.gno (hunk 1 above)
Comment (r3v4s) [reply pool/position.gno:45]: we discussed about making function as method, and this function is already method.

**Review by dongwon8247 — DISMISSED** [review]

File: pool/position.gno [hunk 4]
```diff
@@ -11,12 +11,30 @@ import (
...
+
 // positionGetKey generates a unique key for a position based on the owner's address and the tick range.
 func positionGetKey(
 	owner std.Address,
 	tickLower int32,
 	tickUpper int32,
 ) string {
+	if !owner.IsValid() {
+		panic(addDetailToError(
+			errInvalidAddress,
```
Comment (onlyhyde) [inline pool/position.gno:26]: we already have `errInvalidInput`.
I think that we use this error code when input parameter check error.

File: pool/position.gno [hunk 5]
```diff
@@ -11,12 +11,30 @@ import (
...
+			ufmt.Sprintf("position.gno__positionGetKey() || invalid owner address %s", owner.String()),
+		))
+	}
+
+	if tickLower > tickUpper {
+		panic(addDetailToError(
+			errInvalidTickRange,
+			ufmt.Sprintf("position.gno__positionGetKey() || tickLower(%d) is greater than tickUpper(%d)", tickLower, tickUpper),
```
Comment (onlyhyde) [inline pool/position.gno:34]: This is same with above comment
Error code is `errInvalidInput`. Details are included in error message.
So we know what situation is happend

File: pool/position_test.gno [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		{validAddr, -100, -200, true, `[GNOSWAP-POOL-024] tickLower is greater than tickUpper || position.gno__positionGetKey() || tickLower(-100) is greater than tickUpper(-200)`}, // tickLower > tickUpper
+		{validAddr, 100, 100, false}, // tickLower == tickUpper
+		{validAddr, 100, 200, false}, // tickLower < tickUpper
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { positionGetKey(tc.owner, tc.tickLower, tc.tickUpper) })
+		} else {
+			uassert.NotPanics(t, func() { positionGetKey(tc.owner, tc.tickLower, tc.tickUpper) })
```
Comment (onlyhyde) [inline pool/position_test.gno:38]: Success case has to check value or return data

**Review by notJoon — DISMISSED** [review]

File: pool/position_test.gno [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+		{i256.MustFromDecimal("0"), u256.Zero(), u256.Zero(), true, `[GNOSWAP-POOL-010] zero liquidity || position.gno__positionUpdate() || both liquidityDelta and current position's liquidity are zero`},
+		{i256.MustFromDecimal("100000"), u256.Zero(), u256.Zero(), false, ``},
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { dummyPool.positionUpdateWithKey(positionKey, tc.liquidity, tc.amount0, tc.amount1) })
+		} else {
+			var newPos PositionInfo
+			uassert.NotPanics(t, func() { newPos = dummyPool.positionUpdateWithKey(positionKey, tc.liquidity, tc.amount0, tc.amount1) })
```
Comment (onlyhyde) [inline pool/position_test.gno:79]: positionUpdate tc should check to update liquidity by liquidityDelta.
need to check before and after change value compare

File: pool/position_test.gno [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	}{
+		{i256.MustFromDecimal("0"), u256.Zero(), u256.Zero(), true, `[GNOSWAP-POOL-010] zero liquidity || position.gno__positionUpdate() || both liquidityDelta and current position's liquidity are zero`},
+		{i256.MustFromDecimal("100000"), u256.Zero(), u256.Zero(), false, ``},
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { positionUpdate(position, tc.liquidityDelta, tc.feeGrowthInside0X128, tc.feeGrowthInside1X128) })
+		} else {
+			uassert.NotPanics(t, func() { positionUpdate(position, tc.liquidityDelta, tc.feeGrowthInside0X128, tc.feeGrowthInside1X128) })
```
Comment (onlyhyde) [inline pool/position_test.gno:105]: this is same with above positionUpdateWithKey success case

**Review by onlyhyde — CHANGES_REQUESTED** [review]

File: pool/position_test.gno (hunk 7 above)
Comment (r3v4s) [reply pool/position_test.gno:79]: L#80 checks
https://github.com/gnoswap-labs/gnoswap/blob/7f02c4c274d6da86192c5ca7856bca355e273276/pool/position_test.gno#L80

File: pool/position.gno (hunk 4 above)
Comment (onlyhyde) [reply pool/position.gno:26]: Based on the policy for error wording (using detailed error codes), this code does not need to be modified.

File: pool/position.gno (hunk 5 above)
Comment (onlyhyde) [reply pool/position.gno:34]: Based on the policy for error wording (using detailed error codes), this code does not need to be modified.

File: pool/position_test.gno [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+		{i256.MustFromDecimal("0"), u256.Zero(), u256.Zero(), true, `[GNOSWAP-POOL-010] zero liquidity || position.gno__positionUpdate() || both liquidityDelta and current position's liquidity are zero`, ""},
+		{i256.MustFromDecimal("100000"), u256.Zero(), u256.Zero(), false, "", "100000"},
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { positionUpdate(position, tc.liquidityDelta, tc.feeGrowthInside0X128, tc.feeGrowthInside1X128) })
+		} else {
+			newPos := positionUpdate(position, tc.liquidityDelta, tc.feeGrowthInside0X128, tc.feeGrowthInside1X128)
+			uassert.Equal(t, newPos.liquidity.ToString(), tc.expectedLiquidity)
```
Comment (onlyhyde) [inline pool/position_test.gno:109]: positionUpdate plays two important roles.
1. update liquidity for the position, if there is a liquidity delta
2. calculate and update the accumulated fee for the liquidity provided

There is a check for changing liquidity: 0 -> fluctuating liquidity check.
Non-zero liquidity -> changing liquidity check is also needed to identify any problems in the calculation.
We also need to make sure that the accumulated fee is updated.
Both values are very important in a swap, so it's important to make sure.

**Review by onlyhyde — CHANGES_REQUESTED** [review]

File: pool/position_test.gno (hunk 9 above)
Comment (r3v4s) [reply pool/position_test.gno:142]: a0ce7d1e80554e281b20bbe5996feadaab954f5f

**Review by notJoon — DISMISSED** [review]

File: pool/position_test.gno (hunk 9 above)
Comment (onlyhyde) [reply pool/position_test.gno:142]: @r3v4s 
not enough

positionUpdate set tokensOwed0 and tokensOwed1
we need to check tokensOwed update

Comment (r3v4s) [reply pool/position_test.gno:165]: 66253fc67240b4da9e4e9464351b2fa9cd091a88

**Review by notJoon — DISMISSED** [review]

**Review by onlyhyde — APPROVED** [review]

**Review by notJoon — APPROVED** [review]

**Review by notJoon — APPROVED** [review]

**Review by onlyhyde — APPROVED** [review]

**Pull request merged**